tcp/listener.go

```diff
@@ -0,0 +1,203 @@
+func (s *Server) ListenAndServe(laddr string) (string, error) {
+	if laddr == "" { // Make sure we have an laddr
+		return "", ErrBindAddressRequired
+	}
+
+	ln, err := net.Listen("tcp", laddr)
+	if err != nil {
+		return "", err
+	}
+	s.listener = &ln
+
+	s.Logger.Println("listening on TCP connection", ln.Addr().String())
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		for {
+			// Are we shutting down? If so, exit
+			select {
+			case <-s.shutdown:
+				return
+			default:
+			}
+
+			conn, err := ln.Accept()
+			if opErr, ok := err.(*net.OpError); ok && !opErr.Temporary() {
+				s.Logger.Println("error temporarily accepting TCP connection", err.Error())
+				continue
+			}
+			if err != nil {
+				s.Logger.Println("TCP listener closed")
+				return
+			}
+
+			s.wg.Add(1)
+			go s.handleConnection(conn)
+		}
+	}()
+
+	// Return the host we started up on. Mostly needed for testing
+	return ln.Addr().String(), nil
+}
+
+// Close will close the listener
+func (s *Server) Close() error {
+	// Stop accepting client connections
+	if s.listener != nil {
+		err := (*s.listener).Close()
+		if err != nil {
+			return err
+		}
+	} else {
+		return ErrServerClosed
+	}
+	// Shut down all handlers
+	close(s.shutdown)
+	s.wg.Wait()
+	s.listener = nil
+
+	return nil
+}
+
+// handleConnection services an individual TCP connection.
+func (s *Server) handleConnection(conn net.Conn) {
+	defer func() {
+		conn.Close()
+		s.wg.Done()
+	}()
+
+	messageChannel := make(chan byte)
+
+	// Start our reader up in a go routine so we don't block checking our close channel
+	go func() {
+		var messageType byte
+		err := binary.Read(conn, binary.LittleEndian, &messageType)
+		if err != nil {
+			s.Logger.Printf("unable to read message type %s", err)
+			return
+		}
+		messageChannel <- messageType
+	}()
+
+	for {
+		select {
+		case <-s.shutdown:
+			// Are we shutting down? If so, exit
+			return
+		case messageType := <-messageChannel:
+			switch messageType {
+			case writeShardRequestMessage:
+				err := s.writeShardRequest(conn)
+				s.writeShardResponse(conn, err)
+				return
+			}
+		default:
+		}
+	}
+
+}
+
+func (s *Server) writeShardRequest(conn net.Conn) error {
+	var size int64
+	if err := binary.Read(conn, binary.LittleEndian, &size); err != nil {
+		return err
+	}
+
+	message := make([]byte, size)
+
+	reader := io.LimitReader(conn, size)
+	if _, err := reader.Read(message); err != nil {
+		return err
+	}
+
+	var wsr data.WriteShardRequest
+	if err := wsr.UnmarshalBinary(message); err != nil {
+		return err
+	}
+	if _, err := s.writer.WriteShard(wsr.ShardID(), wsr.Points()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) writeShardResponse(conn net.Conn, e error) {
+	var mt byte = writeShardResponseMessage
+	if err := binary.Write(conn, binary.LittleEndian, &mt); err != nil {
+		s.Logger.Printf("error writing shard response message type: %s", err)
+		return
+	}
+
+	var wsr data.WriteShardResponse
+	if e != nil {
+		wsr.SetCode(1)
```

Comment:
@jwilder I wasn't sure what to do here.  What were you thinking we would do for codes?